geode-core/src/distributedTest/java/org/apache/geode/cache30/CacheStatisticsPartitionedRegionDUnitTest.java

```diff
@@ -0,0 +1,577 @@
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0.0f, rStats.getHitRatio(), 0.0f);
+
+    rStats.resetCounts();
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 0 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0.0f, rStats.getHitRatio(), 0.0f);
+
+    region.put(key, value);
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 0 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0, rStats.getMissCount());
+
+    region.get(key);
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getHitCount());
+    assertEquals(0, rStats.getMissCount());
+    assertEquals(1.0f, rStats.getHitRatio(), 0.0f);
+
+    region.get(key2);
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getMissCount());
+    assertEquals(1, rStats.getHitCount());
+    assertEquals(0.5f, rStats.getHitRatio(), 0.0f);
+
+    region.create(key3, null);
+    region.get(key3); // miss on existing entry
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 2 == rStats.getMissCount());
+    assertEquals(1, rStats.getHitCount());
+    assertEquals(2, rStats.getMissCount());
+    assertEquals(0.33f, rStats.getHitRatio(), 0.01f);
+
+    rStats.resetCounts();
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 0 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0.0f, rStats.getHitRatio(), 0.0f);
+
+    region.invalidate(key);
+    region.get(key);
+
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(1, rStats.getMissCount());
+  }
+
+  /**
+   * Tests that the {@linkplain CacheStatistics#getLastAccessedTime last access time} and
+   * {@link CacheStatistics#getLastModifiedTime last modified time} are updated appropriately for a
+   * partitioned region with only one server.
+   */
+  @Test
+  public void testTimeStats() throws CacheException, InterruptedException {
+    final long ESTAT_RES = 100; // the resolution, in ms, of entry stats
+    String name = this.getUniqueName();
+    Object key = "KEY";
+    Object key2 = "KEY2";
+    Object value = "VALUE";
+    long before;
+    long after;
+    long oldBefore;
+    long oldAfter;
+
+    final int MSECS = 1000;
+
+    AttributesFactory factory = new AttributesFactory();
+    // factory.setScope(Scope.LOCAL);
+    factory.setStatisticsEnabled(true);
+
+    Region region = createPartitionedRegion(name, factory.create());
+    CacheStatistics rStats = region.getStatistics();
+
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    // In order for the stats to have a value, an entry must be accessed first.
+    region.get(key);
```

Comment:
this get is on a key that does not exist. Make that more clear by not using "key". For example "region.get("MISSING_KEY");"
I'm surprised that a get on a non-existent key sets the region's lastAccessed and lastModified.

Reply:
Yes. I think the reason is that when get(missingEntry) is called, the BucketRegion is created and lastAccessedTime and lastModifiedTime are initialized to the current time even if the entry does not exist.
I am surprised about that too.